server/mailrepository/mailrepository-cassandra/src/test/java/org/apache/james/mailrepository/cassandra/CassandraMailRepositoryWithFakeImplementationsTest.java

```diff
@@ -0,0 +1,283 @@
+
+            assertThatThrownBy(() -> cassandraMailRepository.store(mail))
+                    .isInstanceOf(RuntimeException.class)
+                    .hasMessage("java.lang.RuntimeException: Expected failure while saving");
+
+            assertThat(keysDAO.list(URL).join()).isEmpty();
+        }
+    }
+
+    @Nested
+    class FailingMailDaoTest {
+        CassandraMailRepository cassandraMailRepository;
+        CassandraMailRepositoryKeysDAO keysDAO;
+
+        @BeforeEach
+        void setup(CassandraCluster cassandra) {
+            FailingMailDAO mailDAO = new FailingMailDAO(cassandra.getConf(), BLOB_ID_FACTORY, cassandra.getTypesProvider());
+            keysDAO = new CassandraMailRepositoryKeysDAO(cassandra.getConf(), CassandraUtils.WITH_DEFAULT_CONFIGURATION);
+            CassandraMailRepositoryCountDAO countDAO = new CassandraMailRepositoryCountDAO(cassandra.getConf());
+            CassandraBlobsDAO blobsDAO = new CassandraBlobsDAO(cassandra.getConf());
+
+            cassandraMailRepository = new CassandraMailRepository(URL,
+                    keysDAO, countDAO, mailDAO, MimeMessageStore.factory(blobsDAO).mimeMessageStore());
+        }
+
+        class FailingMailDAO extends CassandraMailRepositoryMailDAO {
+
+            public FailingMailDAO(Session session, BlobId.Factory blobIdFactory, CassandraTypesProvider cassandraTypesProvider) {
+                super(session, blobIdFactory, cassandraTypesProvider);
+            }
+
+            @Override
+            public CompletableFuture<Void> store(MailRepositoryUrl url, Mail mail, BlobId headerId, BlobId bodyId) throws MessagingException {
+                return CompletableFuture.supplyAsync(() -> {
+                    throw new RuntimeException("Expected failure while storing mail parts");
+                });
+            }
+        }
+
+        @Test
+        void keysShouldNotBeStoredWhenStoringMailPartsHasFailed() throws Exception {
+            MailKey mailKey = new MailKey("mymail");
+            List<MailAddress> recipients = ImmutableList
+                    .of(new MailAddress("rec1@domain.com"),
+                            new MailAddress("rec2@domain.com"));
+            MimeMessage mailContent = MimeMessageBuilder.mimeMessageBuilder()
+                    .setSubject("test")
+                    .setText("this is the content")
+                    .build();
+            MailImpl mail = new MailImpl(mailKey.asString(), new MailAddress("sender@domain.com"), recipients, mailContent);
+
+            assertThatThrownBy(() -> cassandraMailRepository.store(mail))
+                    .isInstanceOf(RuntimeException.class)
+                    .hasMessage("java.lang.RuntimeException: Expected failure while storing mail parts");
+
+            assertThat(keysDAO.list(URL).join()).isEmpty();
+        }
+
+        @Test
+        void mimeMessageShouldBeStoredWhenStoringMailPartsHasFailed(CassandraCluster cassandra) throws Exception {
```

Comment:
why?

Reply:
IMO this test is not here to enforce a behaviour but demonstate a behaviour.

This should maybe reflect in the test naming:

```
x Is y When z

Here: mimeMessageIsStoredWhenStoringMailPartsHasFailed
```